presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,537 @@
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), outputPositionCount);
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+
+        if (positionCount == outputPositionCount) {
+            Block block = new VariableWidthBlock(positionCount, slice, offsetVector, Optional.ofNullable(includeNulls ? nulls : null));
+            nulls = null;
+            offsetVector = null;
+            data = null;
+            slice = null;
+            return block;
+        }
+
+        boolean[] nullsCopy = null;
+        if (includeNulls) {
+            nullsCopy = new boolean[positionCount];
+        }
+
+        compactValues(positions, positionCount, includeNulls, nullsCopy);
+
+        return new VariableWidthBlock(positionCount, slice, offsetVector, Optional.ofNullable(includeNulls ? nullsCopy : null));
+    }
+
+    private void compactValues(int[] positions, int positionCount, boolean includeNulls, boolean[] nullsCopy)
+    {
+        if (outputPositionsReadOnly) {
+            outputPositions = Arrays.copyOf(outputPositions, outputPositionCount);
+            outputPositionsReadOnly = false;
+        }
+
+        int positionIndex = 0;
+        for (int i = 0; i < outputPositionCount && positionIndex < positionCount; i++) {
+            if (positions[positionIndex] == outputPositions[i]) {
+                if (includeNulls) {
+                    nullsCopy[positionIndex] = nulls[i];
+                }
+                int length = offsetVector[i + 1] - offsetVector[i];
+                if (length > 0) {
+                    System.arraycopy(data, offsetVector[i], data, offsetVector[positionIndex], length);
+                }
+                offsetVector[positionIndex + 1] = offsetVector[positionIndex] + length;
+                positionIndex++;
+            }
+        }
+    }
+
+    @Override
+    public BlockLease getBlockView(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return newLease(new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), outputPositionCount));
+        }
+        boolean includeNulls = nullsAllowed && presentStream != null;
+        compactValues(positions, positionCount, includeNulls, nulls);
```

Comment:
`compactValues` is expensive, don't call it if `positionCount == outputPositionCount`